File: impl/python/agentnet_py/escrow.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from .cbor import CborMap, CborValue


class EscrowError(ValueError):
    pass


@dataclass
class EscrowLockPayload:
    escrow_id: str
    payer: str
    payee: str
    amount: int
    currency: str
    release_condition: CborValue
    dispute_window_sec: int
    expiry: int


@dataclass
class EscrowReleasePayload:
    escrow_id: str
    evidence_receipt_hash: bytes
    ts: int


@dataclass
class EscrowDisputePayload:
    escrow_id: str
    reason: str
    evidence_anchor_or_receipt: bytes
    ts: int


@dataclass
class EscrowResolvePayload:
    escrow_id: str
    outcome: int
    split_amount_to_payee: Optional[int]
    ts: int


def parse_escrow_lock_payload(value: CborValue) -> EscrowLockPayload:
    entries = _expect_map(value)
    return EscrowLockPayload(
        escrow_id=_expect_text(_get_required(entries, 0)),
        payer=_expect_text(_get_required(entries, 1)),
        payee=_expect_text(_get_required(entries, 2)),
        amount=_expect_u64(_get_required(entries, 3)),
        currency=_expect_text(_get_required(entries, 4)),
        release_condition=_get_required(entries, 5),
        dispute_window_sec=_expect_u64(_get_required(entries, 6)),
        expiry=_expect_u64(_get_required(entries, 7)),
    )


def parse_escrow_release_payload(value: CborValue) -> EscrowReleasePayload:
    entries = _expect_map(value)
    evidence = _expect_bytes(_get_required(entries, 1))
    if len(evidence) != 32:
        raise EscrowError("evidence_receipt_hash must be 32 bytes")
    return EscrowReleasePayload(
        escrow_id=_expect_text(_get_required(entries, 0)),
        evidence_receipt_hash=evidence,
        ts=_expect_u64(_get_required(entries, 2)),
    )


def parse_escrow_dispute_payload(value: CborValue) -> EscrowDisputePayload:
    entries = _expect_map(value)
    evidence = _expect_bytes(_get_required(entries, 2))
    if len(evidence) != 32:
        raise EscrowError("evidence_anchor_or_receipt must be 32 bytes")
    return EscrowDisputePayload(
        escrow_id=_expect_text(_get_required(entries, 0)),
        reason=_expect_text(_get_required(entries, 1)),
        evidence_anchor_or_receipt=evidence,
        ts=_expect_u64(_get_required(entries, 3)),
    )


def parse_escrow_resolve_payload(value: CborValue) -> EscrowResolvePayload:
    entries = _expect_map(value)
    split_amount = None
    if _get_optional(entries, 2) is not None:
        split_amount = _expect_u64(_get_optional(entries, 2))
    return EscrowResolvePayload(
        escrow_id=_expect_text(_get_required(entries, 0)),
        outcome=_expect_u8(_get_required(entries, 1)),
        split_amount_to_payee=split_amount,
        ts=_expect_u64(_get_required(entries, 3)),
    )


def escrow_lock_payload_to_cbor(payload: EscrowLockPayload) -> CborValue:
    return CborMap([
        (0, payload.escrow_id),
        (1, payload.payer),
        (2, payload.payee),
        (3, payload.amount),
        (4, payload.currency),
        (5, payload.release_condition),
        (6, payload.dispute_window_sec),
        (7, payload.expiry),
    ])


def escrow_release_payload_to_cbor(payload: EscrowReleasePayload) -> CborValue:
    return CborMap([
        (0, payload.escrow_id),
        (1, payload.evidence_receipt_hash),
        (2, payload.ts),
    ])


def escrow_dispute_payload_to_cbor(payload: EscrowDisputePayload) -> CborValue:
    return CborMap([
        (0, payload.escrow_id),
        (1, payload.reason),
        (2, payload.evidence_anchor_or_receipt),
        (3, payload.ts),
    ])


def escrow_resolve_payload_to_cbor(payload: EscrowResolvePayload) -> CborValue:
    entries = [
        (0, payload.escrow_id),
        (1, payload.outcome),
        (3, payload.ts),
    ]
    if payload.split_amount_to_payee is not None:
        entries.append((2, payload.split_amount_to_payee))
    return CborMap(entries)


def _expect_map(value: CborValue):
    if isinstance(value, CborMap):
        return value.entries
    raise EscrowError("expected map")


def _get_required(entries, key: int) -> CborValue:
    for k, v in entries:
        if isinstance(k, int) and k == key:
            return v
    raise EscrowError("missing required key")


def _get_optional(entries, key: int) -> Optional[CborValue]:
    for k, v in entries:
        if isinstance(k, int) and k == key:
            return v
    return None


def _expect_text(value: CborValue) -> str:
    if isinstance(value, str):
        return value
    raise EscrowError("expected text")


def _expect_bytes(value: CborValue) -> bytes:
    if isinstance(value, (bytes, bytearray)):
        return bytes(value)
    raise EscrowError("expected bytes")


def _expect_u64(value: CborValue) -> int:
    if isinstance(value, int) and value >= 0:
        return value
    raise EscrowError("expected unsigned")


def _expect_u8(value: CborValue) -> int:
    if isinstance(value, int) and 0 <= value <= 0xFF:
        return value
    raise EscrowError("expected u8")

File: impl/ts/src/escrow.ts
import { CborMap, CborValue } from "./cbor.js";

export interface EscrowLockPayload {
  escrowId: string;
  payer: string;
  payee: string;
  amount: number;
  currency: string;
  releaseCondition: CborValue;
  disputeWindowSec: number;
  expiry: number;
}

export interface EscrowReleasePayload {
  escrowId: string;
  evidenceReceiptHash: Uint8Array;
  ts: number;
}

export interface EscrowDisputePayload {
  escrowId: string;
  reason: string;
  evidenceAnchorOrReceipt: Uint8Array;
  ts: number;
}

export interface EscrowResolvePayload {
  escrowId: string;
  outcome: number;
  splitAmountToPayee?: number;
  ts: number;
}

export function parseEscrowLockPayload(value: CborValue): EscrowLockPayload {
  const entries = expectMap(value);
  return {
    escrowId: expectText(getRequired(entries, 0)),
    payer: expectText(getRequired(entries, 1)),
    payee: expectText(getRequired(entries, 2)),
    amount: expectU64(getRequired(entries, 3)),
    currency: expectText(getRequired(entries, 4)),
    releaseCondition: getRequired(entries, 5),
    disputeWindowSec: expectU64(getRequired(entries, 6)),
    expiry: expectU64(getRequired(entries, 7)),
  };
}

export function parseEscrowReleasePayload(value: CborValue): EscrowReleasePayload {
  const entries = expectMap(value);
  const evidence = expectBytes(getRequired(entries, 1));
  if (evidence.length !== 32) {
    throw new Error("evidenceReceiptHash must be 32 bytes");
  }
  return {
    escrowId: expectText(getRequired(entries, 0)),
    evidenceReceiptHash: evidence,
    ts: expectU64(getRequired(entries, 2)),
  };
}

export function parseEscrowDisputePayload(value: CborValue): EscrowDisputePayload {
  const entries = expectMap(value);
  const evidence = expectBytes(getRequired(entries, 2));
  if (evidence.length !== 32) {
    throw new Error("evidenceAnchorOrReceipt must be 32 bytes");
  }
  return {
    escrowId: expectText(getRequired(entries, 0)),
    reason: expectText(getRequired(entries, 1)),
    evidenceAnchorOrReceipt: evidence,
    ts: expectU64(getRequired(entries, 3)),
  };
}

export function parseEscrowResolvePayload(value: CborValue): EscrowResolvePayload {
  const entries = expectMap(value);
  const splitValue = getOptional(entries, 2);
  return {
    escrowId: expectText(getRequired(entries, 0)),
    outcome: expectU8(getRequired(entries, 1)),
    splitAmountToPayee: splitValue === null ? undefined : expectU64(splitValue),
    ts: expectU64(getRequired(entries, 3)),
  };
}

export function escrowLockPayloadToCbor(payload: EscrowLockPayload): CborValue {
  return {
    entries: [
      [0, payload.escrowId],
      [1, payload.payer],
      [2, payload.payee],
      [3, payload.amount],
      [4, payload.currency],
      [5, payload.releaseCondition],
      [6, payload.disputeWindowSec],
      [7, payload.expiry],
    ],
  } satisfies CborMap;
}

export function escrowReleasePayloadToCbor(payload: EscrowReleasePayload): CborValue {
  return {
    entries: [
      [0, payload.escrowId],
      [1, payload.evidenceReceiptHash],
      [2, payload.ts],
    ],
  } satisfies CborMap;
}

export function escrowDisputePayloadToCbor(payload: EscrowDisputePayload): CborValue {
  return {
    entries: [
      [0, payload.escrowId],
      [1, payload.reason],
      [2, payload.evidenceAnchorOrReceipt],
      [3, payload.ts],
    ],
  } satisfies CborMap;
}

export function escrowResolvePayloadToCbor(payload: EscrowResolvePayload): CborValue {
  const entries: [CborValue, CborValue][] = [
    [0, payload.escrowId],
    [1, payload.outcome],
    [3, payload.ts],
  ];
  if (payload.splitAmountToPayee !== undefined) {
    entries.push([2, payload.splitAmountToPayee]);
  }
  return { entries } satisfies CborMap;
}

function expectMap(value: CborValue): [CborValue, CborValue][] {
  if (typeof value === "object" && value !== null && "entries" in value) {
    return (value as CborMap).entries;
  }
  throw new Error("expected map");
}

function getRequired(entries: [CborValue, CborValue][], key: number): CborValue {
  for (const [k, v] of entries) {
    const idx = toNumber(k);
    if (typeof idx === "number" && idx === key) return v;
  }
  throw new Error("missing required key");
}

function getOptional(entries: [CborValue, CborValue][], key: number): CborValue | null {
  for (const [k, v] of entries) {
    const idx = toNumber(k);
    if (typeof idx === "number" && idx === key) return v;
  }
  return null;
}

function expectText(value: CborValue): string {
  if (typeof value === "string") return value;
  throw new Error("expected text");
}

function expectBytes(value: CborValue): Uint8Array {
  if (value instanceof Uint8Array) return value;
  throw new Error("expected bytes");
}

function expectU64(value: CborValue): number {
  if (typeof value === "bigint") {
    const asNumber = Number(value);
    if (!Number.isSafeInteger(asNumber)) {
      throw new Error("expected u64");
    }
    return asNumber;
  }
  if (typeof value === "number" && Number.isSafeInteger(value) && value >= 0) return value;
  throw new Error("expected u64");
}

function expectU8(value: CborValue): number {
  const num = expectU64(value);
  if (num > 0xff) throw new Error("expected u8");
  return num;
}

function toNumber(value: CborValue): number | null {
  if (typeof value === "number") return value;
  if (typeof value === "bigint") {
    const asNumber = Number(value);
    if (!Number.isSafeInteger(asNumber)) return null;
    return asNumber;
  }
  return null;
}
